File: src/api/player/Integration.ts
/**
 * Describes optional playback overrides supplied by a player facade integration.
 *
 * @remarks
 * <br/> - Requires {@link PlayerConfiguration.usePlayerFacade} when creating the player.
 * <br/> - Registration takes place on the Web native handle or through the Android native API, not on the React Native player API.
 * <br/> - Omitted hooks and nullish getter results delegate to the content player.
 * <br/> - The integration owns playback synchronization and must dispatch events when its public state changes.
 *
 * @category Player
 * @platform web,android
 * @public
 */
export interface Integration {
  /**
   * Returns the current playback time, in seconds.
   *
   * @returns A finite, non-negative time, or `null` or `undefined` to use the content player's time.
   * Invalid times also fall back to the content player.
   */
  getCurrentTime?(): number | null | undefined;

  /**
   * Requests a seek to the given playback time.
   *
   * @param currentTime - The requested time, in seconds.
   * @returns `true` if handled, or `false` to seek the content player instead.
   */
  setCurrentTime?(currentTime: number): boolean;

  /**
   * Requests playback to start or resume.
   *
   * @returns `true` if handled, or `false` to play the content player instead.
   */
  play?(): boolean;

  /**
   * Requests playback to pause.
   *
   * @returns `true` if handled, or `false` to pause the content player instead.
   */
  pause?(): boolean;

  /**
   * Returns whether playback is paused.
   *
   * @returns The paused state, or `null` or `undefined` to use the content player's state.
   */
  isPaused?(): boolean | null | undefined;

  /**
   * Returns whether a seek is in progress.
   *
   * @returns The seeking state, or `null` or `undefined` to use the content player's state.
   */
  isSeeking?(): boolean | null | undefined;

  /**
   * Returns the playback duration, in seconds.
   *
   * @returns The duration, or `null` or `undefined` to use the content player's duration.
   * `Infinity` and `NaN` are preserved for live or unknown durations.
   */
  getDuration?(): number | null | undefined;

  /**
   * Returns whether audio is muted.
   *
   * @returns The muted state, or `null` or `undefined` to use the content player's state.
   */
  getMuted?(): boolean | null | undefined;

  /**
   * Requests a change to the muted state.
   *
   * @param muted - Whether audio should be muted.
   * @returns `true` if handled, or `false` to update the content player instead.
   */
  setMuted?(muted: boolean): boolean;

  /**
   * Returns the audio volume.
   *
   * @returns A volume between `0` and `1`, or `null` or `undefined` to use the content player's volume.
   */
  getVolume?(): number | null | undefined;

  /**
   * Requests a change to the audio volume.
   *
   * @param volume - The requested volume, between `0` and `1`.
   * @returns `true` if handled, or `false` to update the content player instead.
   */
  setVolume?(volume: number): boolean;
}

File: android/src/main/java/com/theoplayer/integration/Integration.kt
package com.theoplayer.integration

import androidx.annotation.MainThread
import com.theoplayer.android.api.ads.Ads
import com.theoplayer.android.api.event.ads.AdEvent

/**
 * Optional playback and advertising overrides for [PlayerFacade].
 *
 * Enable `usePlayerFacade` at player creation and register through [com.theoplayer.ReactTHEOplayerView.registerIntegration].
 * Unimplemented hooks delegate to the content player. The integration owns playback synchronization,
 * advertising lifecycle, and events describing its public state. All hooks run on the main thread.
 */
@MainThread
interface Integration {
  /** Returns a finite, non-negative time in seconds, or null to delegate. Invalid times also delegate. */
  fun getCurrentTime(): Double? = null

  /** Seeks to [currentTime] in seconds. Returns true if handled, or false to seek the content player. */
  fun setCurrentTime(currentTime: Double): Boolean = false

  /** Starts or resumes playback. Returns true if handled, or false to play the content player. */
  fun play(): Boolean = false

  /** Pauses playback. Returns true if handled, or false to pause the content player. */
  fun pause(): Boolean = false

  /** Returns the paused state, or null to use the content player's state. */
  fun isPaused(): Boolean? = null

  /** Returns the seeking state, or null to use the content player's state. */
  fun isSeeking(): Boolean? = null

  /** Returns the duration in seconds, or null to delegate. Infinity and NaN are preserved. */
  fun getDuration(): Double? = null

  /** Returns the muted state, or null to use the content player's state. */
  fun getMuted(): Boolean? = null

  /** Sets [muted]. Returns true if handled, or false to update the content player. */
  fun setMuted(muted: Boolean): Boolean = false

  /** Returns the volume between 0 and 1, or null to use the content player's volume. */
  fun getVolume(): Double? = null

  /** Sets [volume] between 0 and 1. Returns true if handled, or false to update the content player. */
  fun setVolume(volume: Double): Boolean = false

  /**
   * The Ads API and event source behind the stable facade Ads object, or null to retain native Ads.
   * Keep this object stable until registration closes. Existing listeners migrate on registration and close.
   */
  val ads: Ads? get() = null

  /**
   * Returns advertising state without replacing the native Ads API, or null to delegate.
   * Ignored when [ads] is supplied. State is never derived from dispatched events.
   */
  fun getAdState(): PlayerFacadeAdState? = null

  /**
   * Returns true to consume a native Ads [event] for all facade listeners, or false to forward it.
   * Called once per subscribed event. Integration-generated events bypass this hook.
   */
  fun shouldConsumeAdEvent(event: AdEvent<*>): Boolean = false
}
